### DISCUSSION TITLE
Utility classes not found warning on build

### DISCUSSION THREAD

**Discussion by col-bc** [tailwindlabs/tailwindcss#7621] [Help] [answered]
I am having trouble building the css with the CLI tool. I get a warning when building the css about the `content` option in `tailwind.config.js`. The build results in a very bare css file without the base, components, or utilities classes.
```
$ npx tailwindcss -i ./src/tailwind-in.css -o ./src/assets/global.css

warn - No utility classes were detected in your source files. If this is unexpected, double-check the `content` option in your Tailwind CSS configuration.
warn - https://tailwindcss.com/docs/content-configuration

Done in 227ms.
```

I have a suspicion that the config file or `npx` command are having problems finding my css file with the `@tailwind ...` directives. When I change `content` to `content: ["./**/*.{html,js}"],` (without the leading `./src`), tailwind compiles without the warning and produces a much larger output file. However, I get the following runtime error:

File structure:
```
├── node_modules
├── package-lock.json
├── package.json
├── public
│   ├── favicon.ico
│   └── index.html
├── src
│   ├── App.vue
│   ├── assets
│   │   ├── global.css
│   │   ├── logo.png
│   │   └── tailwind.css
│   ├── components
│   │   └── HelloWorld.vue
│   ├── main.ts
│   ├── router
│   │   └── index.ts
│   ├── shims-vue.d.ts
│   ├── store
│   │   └── index.ts
│   ├── tailwind-in.css
│   └── views
│       ├── About.vue
│       └── Home.vue
├── tailwind.config.js
└── tsconfig.json
```

tailwindcss.config.js
```
module.exports = {
  content: ["./src/**/*.{html,js}"],
  theme: {
    extend: {},
  },
  plugins: [],
}
```

./src/tailwind-in.css
```
@tailwind base;
@tailwind components;
@tailwind utilities;
```

I am very lost. I'd appreciate any assistance available. 

Edit
Added leading slash to `content` in `tailwindcss.config.js` content to match the code causing this issue.

**Comment by adamwathan** [comment]
Hey! Your content path is wrong, you’re missing a slash:

```diff
  module.exports = {
-   content: [".src/**/*.{html,js}"],
+   content: ["./src/**/*.{html,js}"],
    theme: {
      extend: {},
    },
    plugins: [],
  }
```

**Reply by col-bc** [reply]
My mistake, that was a typo on this discussion. My config includes the leading slash.

**Comment by col-bc** [comment, marked as answer]
I was able to resolve the problem. Here is the working code.
```
content: [
    "./src/**/*.{html, js, ts, vue}", 
    "./src/**/*"
],
```

**Reply by exjade** [reply]
Thanks!

**Reply by angelhdzmultimedia** [reply]
> "./src/**/*.{html, js, ts, vue}", 
>     "./src/**/*"

I'm using this solution and it doesn't work for me.
I have a Fresh Vite + Vue 3 + TypeScript project.
Tailwind works but I can't get rid of that warning...
I've searched a lot, in Github issues, in Google...

Any help will be appreciated, I'm making a tutorial on YouTube in how to add TailwindCSS to a Vite project and faced this wall.
For me it's just a warning I can ignore it, but for the future viewers it can discourage them.

**Reply by angelhdzmultimedia** [reply]
> > "./src/**/*.{html, js, ts, vue}",
> > "./src/**/*"
> 
> I'm using this solution and it doesn't work for me. I have a Fresh Vite + Vue 3 + TypeScript project. Tailwind works but I can't get rid of that warning... I've searched a lot, in Github issues, in Google...
> 
> Any help will be appreciated, I'm making a tutorial on YouTube in how to add TailwindCSS to a Vite project and faced this wall. For me it's just a warning I can ignore it, but for the future viewers it can discourage them.

I think I kinda solved it?
Thanks to someone on Discord.
I removed `@tailwind utilities;` from my `./src/index.css` and the warning went away. 

Now I'm wondering if there's a real solution as how to define those "utilities" so it doesn't fail to find them...?

**Reply by col-bc** [reply]
After using tailwind in a few more projects, I have discovered that this surfaces frequently when using `vue-router` and the solution does not work without explicitly listing the `views` directory.
```
content: [
  './public/index.html',
  './src/**/*.{vue,js,ts}',
  './src/views/**/*.{vue,js,ts}',
  './src/components/**/*.{vue,js,ts}',
]
```

I have also found that I can put the `@tailwind` directives directly into my `App.vue` file. This eliminates a file and import, making it slightly easier for me to manage.

```
// App.vue
<template>
  <router-view />
</template>

<style>
@tailwind base;
@tailwind components;
@tailwind utilities;
@tailwind variants;
</style>

**Reply by angelhdzmultimedia** [reply]
> After using tailwind in a few more projects, I have discovered that this surfaces frequently when using `vue-router` and the solution does not work without explicitly listing the `views` directory.
> 
> ```
> content: [
>   './public/index.html',
>   './src/**/*.{vue,js,ts}',
>   './src/views/**/*.{vue,js,ts}',
>   './src/components/**/*.{vue,js,ts}',
> ]
> ```
> 
> I have also found that I can put the `@tailwind` directives directly into my `App.vue` file. This eliminates a file and import, making it slightly easier for me to manage.
> 
> ```
> // App.vue
> <template>
>   <router-view />
> </template>
> 
> <style>
> @tailwind base;
> @tailwind components;
> @tailwind utilities;
> @tailwind variants;
> </style>
> ```

### Solved

Thank you. Thank you. Thank you! Finally solved it.
And I will try to tell many people about this so they don't have to struggle like I did.

For me, it was happening exactly what the warning was saying I wasn't using any utility classes, because in my specific case, I'm building a video Tutorial for YouTube step-by-step  to help people install TailwindCSS in a Vite/Vue 3/TypeScript project.

So I was at a step where I wasn't adding any utility classes yet, so obviously it will warn me!

And the reason I didn't get this warning the first time I tried Tailwind is because I did added an utility class!

The moment I added `text-blue-500` on a `<h1>`, the warn went away!!! 

Wishing you health, joy and success!

**Reply by col-bc** [reply]
@angelhdzmultimedia, happy to help. Also check out [this vue cli plugin](https://github.com/forsartis/vue-cli-plugin-tailwind) for an even easier setup. Your viewers may appreciate it :)

**Reply by Aitizazhassan** [reply]
> I was able to resolve the problem. Here is the working code.
> 
> ```
> content: [
>     "./src/**/*.{html, js, ts, vue}", 
>     "./src/**/*"
> ],
> ```

This solution work for me thanks

**Reply by funylon** [reply]
In my case it was a relative path issue. 

I'm working under the webserver's docroot htdocs/myproject. 

In the twcli docs example src/index.html:
 <link href="/dist/output.css" rel="stylesheet"> needed to be 
 <link href="../dist/output.css" rel="stylesheet">

Now the right dist/output.css is used for:
npx tailwindcss -i ./src/input.css -o ./dist/output.css --watch

**Comment by akashpathak9587** [comment]
this error is just because of
content: [
],
instead do this
content: ["./src/**/*.{html,js}"],

And classes get automatically added in output file as soon as you will use them in html file.
Just follow the installation setp as on the telwind website  https://tailwindcss.com/docs/installation

**Comment by mxvsh** [comment]
I tried all the above, but none worked. I had to use `path` in the `tailwind.config.js` file to make it work.

```js
const { join } = require('path');

module.exports = {
  content: [join(__dirname, 'src/**/*.{js,ts,jsx,tsx}')],
  theme: {
    extend: {},
  },
  plugins: [],
};

```

**Reply by PetrSasha** [reply]
thx, works

**Reply by jaskin21** [reply]
thx bro. it works

**Reply by faisaldeveloped** [reply]
Thank you bro :)

**Reply by nestorbfgo** [reply]
i don´t know why but it's working, thx!

**Reply by ssadek-coder** [reply]
Spent hours on this issue and this is the only solution that worked. Thank you!

**Reply by Gauravkumario** [reply]
thank you bro for solution

**Reply by luchillo17** [reply]
This seem to happen a lot with monorepos like Nx where the root of the repo is not the root of the project/app, I also had to path join the dirname of the project.
```mjs
import tailwind from '@astrojs/tailwind';
import { defineConfig } from 'astro/config';
import { join } from 'node:path';

// https://astro.build/config
export default defineConfig({
  outDir: '../../dist/apps/portfolio',
  integrations: [
    tailwind({
      configFile: join(import.meta.dirname, './tailwind.config.mjs'),
    }),
  ],
});
```

**Reply by luchillo17** [reply]
And had to do the same for the `taiwind.config.mjs`:
```mjs
import { join } from 'node:path';

/** @type {import('tailwindcss').Config} */
export default {
  content: [
    join(__dirname, './src/**/*.{astro,html,js,jsx,md,mdx,svelte,ts,tsx,vue}'),
  ],
  theme: {
    extend: {},
  },
  plugins: [],
};

```

**Comment by Milleruriol** [comment]
The content section of your tailwind.config.js file is where you configure the paths to all of your HTML templates, JavaScript components, and any other source files that contain Tailwind class names.

module.exports = {
  content: [
    './pages/**/*.{html,js}',
    './components/**/*.{html,js}',
  ],
  theme: {
    extend: {},
  },
  plugins: [],
}

**Comment by Voyager-Two** [comment]
Mine is only broken in production `NODE_ENV=production`, while using Next.js + Docker + DigitalOcean app platform.
I can't seem to find a fix for it. I've tried it with tailwindcss package as non-dev dependency.

I wonder if it's related to Next.js `cache` folder somehow. As when I update `tailwind.config.js`, I have to clear Next.js cache.
Also for some reason, it seems to only detect tailwind `content` inside `postcss.config.js`, instead of  `tailwind.config.js`.

Edit: it was because I had an unusual setup with Next.js being served through Nest.js

**Comment by webianasan** [comment]
Make sure, you place both the index.html and input.css file in the src file as instructed in the installation guidelines.
I made this mistake of keeping the two files in two different folder and my project didn't work. After fixing that it is working.

**Reply by Jessica-Ryan** [reply]
This works

**Reply by Jessica-Ryan** [reply]
Only issue is that vite can't find html file...? Get error: This localhost page can’t be found

**Reply by Ran-sy** [reply]
add /src/index.html

**Reply by santoshdhonde** [reply]
Yes, It's works, I have also same problem, used wrong path for index.html

**Comment by aifodu** [comment]
In my case, I has everything setup but had not used any tailwind class. I tried the aforementioned with no success. But once I used a tailwind class, all was well.

**Comment by Selikeh** [comment]
The solution is actually quite simple. Your content paths should point to your HTML files that make use of the tailwind utility classes. You could either move your HTML files to the src folder or add paths to your HTML files in the content section like so:
```javascript

/** @type {import('tailwindcss').Config} */
module.exports = {
  content: ["./src/**/*.{html,js}",
            "./*.{html,js}"],
  theme: {
    extend: {},
  },
  plugins: [],
}

```
The second path in the content section above points to HTML and JS files in the main parent folder. Also, your HTML file should contain some utility classes otherwise the error would persist, so throw some utility classes into your HTML to prevent the error from occurring. Don't forget to link your output CSS file to the HTML file.

**Reply by ab1820** [reply]
This worked for me! as of January 2022. So anyone having issues with the config file and styles not working, use this! Thanks @Selikeh

**Comment by ManfredHu** [comment]
maybe document should be update https://tailwindcss.com/docs/guides/vite#vue 🤣

**Comment by n4j1Br4ch1D** [comment]
In my case I had my project in sub directories I solved like this:

//tailwind.config.js
```js
content: ["./src/**/*.{htm,html,js}"],
```
// package.json
```json
  "scripts": {
    "tailwindcss": "cd test/demo && npx tailwindcss -i ./src/css/input.css -o ./src/build/output.css --watch",
  },
```

cd test/demo && npx tailwindcss -i ./src/css/input.css -o ./src/css/output.css --watch

**Comment by PANxxj** [comment]
None of above was working in my project so i Try this and it worked
/** @type {import('tailwindcss').Config} */
export default {
  content: [
    "./index.html",
    "./src/**/*.{vue,js,ts,jsx,tsx}",
  ],
  theme: {
    extend: {},
  },
  plugins: [],
}

**Comment by MrDBC** [comment]
I'm working on a nextjs project( I'm new to it)
and it has no index.html file

I don't have a src folder,
the pages folder is there, 
```
/** @type {import('tailwindcss').Config} */

const { join } = require('path');

module.exports = {
  content: [
    'pages/**/*.{js,ts,jsx,tsx}',
    'components/**/*.{js,ts,jsx,tsx}',
  ],
  theme: {
    extend: {},
  },
  plugins: [],
}


```

attaching folder/file structure
![image](https://github.com/tailwindlabs/tailwindcss/assets/69349350/88353f69-e0df-4ba9-b5fc-5351f01f4f52)


what is the solution for that?


Edit: sorry guys, the problem was in the globals.css file:

```
@tailwind base;
@tailwind components;
@tailwind utilities;

.root{
    @apply h-full bg-black;
}

.body{
    @apply h-full bg-black;
}
```

instead of :

```
@tailwind base;
@tailwind components;
@tailwind utilities;

:root{
    @apply h-full bg-black; 
}

body{
    @apply h-full bg-black;
}
```

LOL

**Comment by BESSETTY-AKHIL** [comment]
If there are no Tailwind utility classes in your files, Tailwind will generate this warning. Add some basic utility classes to test:
apply this code to the (App.tsx)



````import React from 'react';

const App: React.FC = () => {
  return (
    <div className="flex items-center justify-center min-h-screen bg-gray-100">
      <h1 className="text-3xl font-bold text-blue-500">Hello, Tailwind!</h1>
    </div>
  );
};

export default App;

**Comment by Timi-Jegede** [comment]
content: ["./src/**/*.{html,js}",
            "./*.{html,js}"],
Using this worked for me.
